src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitService.kt

```diff
@@ -0,0 +1,169 @@
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.Visit
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitOrderAdjustmentReason
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitStatus
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitType
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.VisitVisitor
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.AgencyLocationRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderBookingRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderVisitBalanceAdjustmentRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.OffenderVisitBalanceRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.ReferenceCodeRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.VisitRepository
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.repository.VisitVisitorRepository
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.util.function.Supplier
+
+@Service
+@Transactional
+class VisitService(
+  private val visitRepository: VisitRepository,
+  private val visitVisitorRepository: VisitVisitorRepository,
+  private val offenderBookingRepository: OffenderBookingRepository,
+  private val offenderVisitBalanceRepository: OffenderVisitBalanceRepository,
+  private val offenderVisitBalanceAdjustmentRepository: OffenderVisitBalanceAdjustmentRepository,
+  private val eventStatusRepository: ReferenceCodeRepository<EventStatus>,
+  private val visitTypeRepository: ReferenceCodeRepository<VisitType>,
+  private val visitStatusRepository: ReferenceCodeRepository<VisitStatus>,
+  private val agencyRepository: AgencyLocationRepository,
+  private val telemetryClient: TelemetryClient,
+) {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun createVisit(visitDto: CreateVisitRequest): CreateVisitResponse {
+
+    val offenderBooking = offenderBookingRepository.findByOffenderNomsIdAndActive(visitDto.offenderNo, true)
+      .orElseThrow(PrisonerNotFoundException(visitDto.offenderNo))
+
+    createBalance(visitDto, offenderBooking)
+
+    val visit = visitRepository.save(mapVisitModel(visitDto, offenderBooking))
+
+    addVisitors(visit, offenderBooking, visitDto)
+
+    telemetryClient.trackEvent("visit-created", mapOf("visitId" to visit.id!!.toString()), null)
+
+    return CreateVisitResponse(visit.id)
+  }
+
+  fun amendVisit(visitId: Long, visitDto: AmendVisitRequest) {
+  }
+
+  fun cancelVisit(visitId: Long, visitDto: CancelVisitRequest) {
+  }
+
+  private fun createBalance(
+    visitDto: CreateVisitRequest,
+    offenderBooking: OffenderBooking,
+  ) {
+    val offenderVisitBalance = offenderVisitBalanceRepository.findById(offenderBooking.bookingId!!).orElseThrow()
```

Comment:
`bookingId!!`  implies it is optional? Is could we make it not optional or use `lateinit` it feels there should be away around describing an id as mandatory but it is set as soon as the entity is created.
Not sure what we have done elsewhere

Reply:
id: Long = 0 🙂